clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1652,6 +1652,16 @@ void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
       constructKernelName(*this, KernelCallerFunc, MC);
   StringRef KernelName(getLangOpts().SYCLUnnamedLambda ? StableName
                                                        : CalculatedName);
+  if (KernelLambda->isLambda()) {
+    llvm::DenseMap<const VarDecl *, FieldDecl *> Captures;
+    FieldDecl *ThisCapture;
+    KernelLambda->getCaptureFields(Captures, ThisCapture);
+    for (const LambdaCapture &LC : KernelLambda->captures()) {
+      if (LC.capturesThis() && LC.isImplicit()) {
+        Diag(LC.getLocation(), diag::err_implicit_this_capture);
+      }
+    }
+  }
```

Comment:
Eeh, this is a bit outside of the whole Visitor model that we have in SemaSYCL. We often use `SyclKernelFieldChecker ` to check kernel fields. The problem that it actually checks **fields** not captures. Right now, I don't know how to integrate it into the current visitor model, though.
@elizabethandrews , @erichkeane , any ideas?

Reply:
Yeah, theres not really a good way to do this with the field checker.  If this is a limitation in the standard, it is definitely something that should be diagnosed.

However, this message is wrong I think, this also looks like it will diagnose on capture with '&' of this as well.

Also, 1656,57, and 58 are unnecessary as far as I can tell, those lines likely need to be removed.

Reply:
Actually SYCL 1.2.1 standard prohibits all sorts of passing pointers from host to device, however we support USM extension that actually allows some pointers to be passed to the kernel, but `this` pointer is definitely not the case of USM pointer and accidental capture of `this` pointer is very common in user's code and always produces fails in run time. This definitely should be diagnosed.

Reply:
> Actually SYCL 1.2.1 standard prohibits all sorts of passing pointers from host to device, however we support USM extension that actually allows some pointers to be passed to the kernel, but `this` pointer is definitely not the case of USM pointer and accidental capture of `this` pointer is very common in user's code and always produces fails in run time. This definitely should be diagnosed.

I think this should be an error only if strict compliance to SYCL 1.2.1 is requested; if USM is enabled, it should be a warning instead.

My understanding is that if a class is allocated in USM, the `this` pointer _is_ a USM pointer and can be safely used in kernels.

Reply:
Hi @Pennycook , Thanks for the comment. I agree that we should deal with USM pointers as a special case. Is there a way that the compiler can know if USM is enabled? Also, for a given class object, is it possible for the compiler to check if it is allocated in USM? I tried looking around in compiler sources and I did not see any other instance where we are checking for USM. May be I missed something.

Reply:
@erichkeane. Thanks for your comment. I will remove 1656-58. I was using them earlier. Also, I will check to see if using [&] results in similar runtime error and update the error message accordingly.

Reply:
> Is there a way that the compiler can know if USM is enabled?

Adding @jbrodman and @AlexeySachkov.  I seem to remember talk of providing something like an `-fsycl-usm` flag to control the USM extension, but I don't know what it's implementation status is.

> Also, for a given class object, is it possible for the compiler to check if it is allocated in USM?

Unfortunately not.  Whether the object was allocated in USM or not is something that can only be determined at runtime.  This is why I suggested that we generate a warning in this case; a user would then see at compile-time that what they're doing _may_ be unsafe, but would be free to ignore the warning if they wanted.

Alternatively, if we decide to stick with an error for implicit capture of `this` in all situations, we could display different error messages.  If USM is disabled, capture of `this` is not allowed for SYCL kernels.  If USM is enabled, capture of `this` is only allowed if the class was allocated in USM, and the user should explicitly specify `this` in the capture list if it was intended.

Reply:
Hi @Pennycook 
Thanks so much for the detailed feedback. I agree that we can emit two different error messages. But, I think that can be captured as a separate task. Please advise.

Reply:
I think 'strict-sycl' (or sycl-strict?) was intended for this type of extension.  I think we intended to have pointers in general handled with that flag.

In 'strict' SYCL mode, ALL captures of pointers (including 'this') is supposed to be ill-formed, by my understanding, so the extra rule of 'implicit capture of this' seems unnecessary*.  That said, having our USM extension require explicit capture seems like a good idea, accidental capture of 'this' is unfortunately quite easy.

*The exception of course is capture via '=', where someone who doesn't understand how implicit this capture works would be burned.  Since '=' means "take the state as it is now", accidental capture of 'this' is particularly surprising/dangerous.

Reply:
Yes, accidental capture of this through [=] has caused a lot of pain.